streaming-payments.md

```diff
@@ -0,0 +1,375 @@
+
+function sendTransfer (obj) {
+  obj.id = uuid()
+  obj.from = plugin.getAccount()
+  // to
+  obj.ledger = plugin.getInfo().prefix
+  // amount
+  // executionCondition
+  obj.expiresAt = new Date(new Date().getTime() + 1000000).toISOString()
+  return plugin.sendTransfer(obj).then(function () {
+    return obj.id
+  })
+}
+
+plugin.connect().then(function () {
+  return fetch('http://localhost:8000/')
+}).then(function (inRes) {
+  inRes.body.pipe(process.stdout)
+  const payHeaderParts = inRes.headers.get('Pay').split(' ')
+  console.log(payHeaderParts)
+  // e.g. Pay: 1 test.crypto.xrp.asdfaqefq3f.26wrgevaew SkTcFTZCBKgP6A6QOUVcwWCCgYIP4rJPHlIzreavHdU
+  setInterval(function () {
+    const ilpPacketContents = {
+      account: payHeaderParts[1] + '.' + (++counter),
+      amount: '1',
+      data: ''
+    }
+    const fulfillment = hmac(Buffer.from(payHeaderParts[2], 'base64'), ilpPacketContents.account)
+    const condition = hash(fulfillment)
+    sendTransfer({
+      to: payHeaderParts[1],
+      amount: '1',
+      executionCondition: base64(condition),
+      ilp: base64(IlpPacket.serializeIlpPayment(ilpPacketContents))
+    }).then(function () {
+      // console.log('transfer sent')
+    }, function (err) {
+      console.error(err.message)
+    })
+  }, 500)
+})
+```
+
+As you can see, it parses the `Pay` header, and then sends one XRP per 500ms. Try it out!
+
+```sh
+$ node ./tool.js
+```
+
+After some startup time, you should see letters being printed.
+
+## Using a trustline
+
+Getting one letter per 500ms is not very fast. It would be nice if we could stream the money faster, so that
+the content arrives faster! For this, we can add a ledger to the shop. The tool opens an account on this ledger,
+and then pays for letters from its account at the shop's ledger, which will be much faster
+than paying via the XRP ledger.
+
+The shop's ledger will expose the Bilateral Transfer Protocol (BTP), which is an optimization of the Ledger Plugin Interface (LPI)
+that we already saw in the Letter Shop tutorial. It is symmetrical over a WebSocket, so the client can send the same messages as the
+server, and this will have the same effect, but in the opposite direction. For instance, both parties can send a transfer to the
+other party in the form of a 'PREPARE' packet, and both parties can also fulfill transfers from the other party by sending a
+'FULFILL' packet. These BTP packets are similar to the objects passed to `plugin.sendTransfer` or `plugin.fulfillCondition`,
+although they are a bit more concise, and before they go onto the WebSocket, they are serialized into OER buffers. To learn
+more about the BTP protocol, read [the BTP spec](https://github.com/interledger/rfcs/pull/300).
+
+Thanks to the plugin architeture, we have to change surprisingly little to switch from XRP to BTP: we just include the
+`'ilp-plugin-payment-channel-framework'` plugin instead of the `'ilp-plugin-xrp-escrow'` one, and give it the config options
```

Comment:
We should probably make a super slimmed down version of this that only implements the client side of BTP. It's strange to include a module with that title here

Reply:
Yes! That's what https://github.com/interledgerjs/btp-toolbox/pull/1 and https://github.com/interledgerjs/btp-toolbox/pull/2 were for, and also https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/issues/31. Let's discuss that option in our 1:1 today.